plugin/template/README.md

```diff
@@ -0,0 +1,166 @@
+# template
+
+*template* - allows for dynamic responses based on the incoming query.
+
+## Description
+
+The *template* plugin allows you to dynamically repond to queries by just writing a (Go) template.
+
+## Syntax
+
+~~~
+template CLASS TYPE [REGEX...] {
+    [answer RR]
+    [answer RR]
+    [additional RR]
+    [...]
+    [rcode responsecode]
+}
+~~~
+
+* **CLASS** the query class (usually IN or ANY)
+* **TYPE** the query type (A, PTR, ...)
+* **REGEX** [Go regexp](https://golang.org/pkg/regexp/) that are matched against the incoming query. Specifying no regex matches everything (default: `.*`). First matching regex wins.
+* `RR` A [RFC 1035](https://tools.ietf.org/html/rfc1035#section-5) style `<rr>` fragment build by a [Go template](https://golang.org/pkg/text/template/) that contains the answer.
+* `responsecode` A response code (`NXDOMAIN, SERVFAIL, ...`). The default is `SUCCESS`.
+
+At least one answer section or rcode is needed.
+
+[Also see](#also-see) contains an additional reading list.
+
+## Templates
+
+Each resource record is a full-featured [Go template](https://golang.org/pkg/text/template/) with the following predefined data
+* `.Name` the query name, as a string
+* `.Class` the query class (usually `IN`)
+* `.Type` the RR type requested (e.g. `PTR`)
+* `.Match` an array of all matches. `index .Match 0` refers to the whole match.
+* `.Group` a map of the named capture groups.
+* `.Message` the incoming DNS query message.
+* `.Question` the matched question section.
+
+The output of the template must be a [RFC 1035](https://tools.ietf.org/html/rfc1035) style resource record line (commonly refered to as a "zone file").
+
+**WARNING** there is a syntactical problem with Go templates and caddy config files. Expressions like `{{$var}}` will be interpreted as a reference to an environment variable by caddy/coredns while `{{ $var }}` will work. Try to avoid template variables. See [Bugs](#bugs)
+
+## Metrics
+
+If monitoring is enabled (via the *prometheus* directive) then the following metrics are exported:
+- `coredns_template_matches_total` the total number of matched requests.
+- `coredns_template_template_failures_total` the number of times the Go templating failed.
+- `coredns_template_rr_failures_total` the number of times the templated resource record was invalid and could not be parsed.
+
+Both failure cases indicate a problem with the template configuration.
```

Comment:
Are these runtime errors? I.e. can we catch the template config errors on startup?

Reply:
Yes, even if the template is syntactical correct it's hard to tell if
1. it works with the input
2. the output is a valid RR

Regex and template syntax errors will break startup, though.